tapchannelmsg/records.go

```diff
@@ -0,0 +1,1568 @@
+	if err != nil {
+		return err
+	}
+
+	if _, ok := tlvs[auxLeaf.TlvType()]; ok {
+		h.AuxLeaf = tlv.SomeRecordT(auxLeaf)
+	}
+
+	if _, ok := tlvs[secondLevelLeaf.TlvType()]; ok {
+		h.SecondLevelLeaf = tlv.SomeRecordT(secondLevelLeaf)
+	}
+
+	return nil
+}
+
+// Bytes returns the serialized HtlcAuxLeaf record.
+func (h *HtlcAuxLeaf) Bytes() []byte {
+	var buf bytes.Buffer
+	_ = h.Encode(&buf)
+	return buf.Bytes()
+}
+
+// DecodeHtlcAuxLeaf deserializes a HtlcAuxLeaf from the given blob.
+func DecodeHtlcAuxLeaf(blob tlv.Blob) (*HtlcAuxLeaf, error) {
+	var h HtlcAuxLeaf
+	err := h.Decode(bytes.NewReader(blob))
+	if err != nil {
+		return nil, err
+	}
+
+	return &h, nil
+}
+
+// HtlcAuxLeafMapRecord is a record that represents a map of HTLC indices to
+// HtlcAuxLeaf records.
+type HtlcAuxLeafMapRecord struct {
+	HtlcAuxLeaves map[input.HtlcIndex]HtlcAuxLeaf
+}
+
+// NewHtlcAuxLeafMapRecord creates a new HtlcAuxLeafMapRecord record with the
+// given HTLC aux leaves.
+func NewHtlcAuxLeafMapRecord(
+	leaves map[input.HtlcIndex]input.HtlcAuxLeaf) HtlcAuxLeafMapRecord {
+
+	if leaves == nil {
+		return HtlcAuxLeafMapRecord{}
+	}
+
+	htlcLeaves := make(map[input.HtlcIndex]HtlcAuxLeaf)
+	for htlcIndex := range leaves {
+		htlcLeaves[htlcIndex] = NewHtlcAuxLeaf(leaves[htlcIndex])
+	}
+
+	return HtlcAuxLeafMapRecord{
+		HtlcAuxLeaves: htlcLeaves,
+	}
+}
+
+// Record creates a Record out of a HtlcAuxLeafMapRecord using the
+// eHtlcAuxLeafMapRecord and dHtlcAuxLeafMapRecord functions.
+//
+// NOTE: This is part of the tlv.RecordProducer interface.
+func (l *HtlcAuxLeafMapRecord) Record() tlv.Record {
+	size := func() uint64 {
+		var (
+			buf     bytes.Buffer
+			scratch [8]byte
+		)
+		err := eHtlcAuxLeafMapRecord(&buf, &l.HtlcAuxLeaves, &scratch)
+		if err != nil {
+			panic(err)
+		}
+
+		return uint64(buf.Len())
+	}
+
+	// Note that we set the type here as zero, as when used with a
+	// tlv.RecordT, the type param will be used as the type.
+	return tlv.MakeDynamicRecord(
+		0, &l.HtlcAuxLeaves, size, eHtlcAuxLeafMapRecord,
+		dHtlcAuxLeafMapRecord,
+	)
+}
+
+// Encode serializes the htlcPartialSigsRecord to the given io.Writer.
+func (l *HtlcAuxLeafMapRecord) Encode(w io.Writer) error {
+	// Create the tlv stream.
+	tlvStream, err := tlv.NewStream(l.Record())
+	if err != nil {
+		return err
+	}
+
+	return tlvStream.Encode(w)
+}
+
+// Decode deserializes the htlcPartialSigsRecord from the given io.Reader.
+func (l *HtlcAuxLeafMapRecord) Decode(r io.Reader) error {
+	// Create the tlv stream.
+	tlvStream, err := tlv.NewStream(l.Record())
+	if err != nil {
+		return err
+	}
+
+	return tlvStream.Decode(r)
+}
+
+// eHtlcAuxLeafMapRecord is an encoder for HtlcAuxLeafMapRecord.
+func eHtlcAuxLeafMapRecord(w io.Writer, val interface{}, buf *[8]byte) error {
+	if v, ok := val.(*map[input.HtlcIndex]HtlcAuxLeaf); ok {
+		numHtlcs := uint64(len(*v))
+		if err := tlv.WriteVarInt(w, numHtlcs, buf); err != nil {
+			return err
+		}
```

Comment:
The max check when decoding is to avoid a memory allocation DoS, since we decode stuff sent to us over the wire. I don't think we need to defend against us creating too many HTLCs, at least not at this (wire) level.